flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTask.java

```diff
@@ -114,4 +134,67 @@ protected void createInputProcessor(
 			operatorChain,
 			setupNumRecordsInCounter(mainOperator));
 	}
+
+	@Override
+	public Future<Boolean> triggerCheckpointAsync(
+			CheckpointMetaData metadata,
+			CheckpointOptions options,
+			boolean advanceToEndOfEventTime) {
+
+		CompletableFuture<Boolean> resultFuture = new CompletableFuture<>();
+		mainMailboxExecutor.execute(
+			() -> {
+				try {
+					pendingCheckpointCompletedFutures.put(metadata.getCheckpointId(), resultFuture);
+					triggerSourcesCheckpoint(new CheckpointBarrier(metadata.getCheckpointId(), metadata.getTimestamp(), options));
+				}
+				catch (Exception ex) {
+					// Report the failure both via the Future result but also to the mailbox
+					pendingCheckpointCompletedFutures.remove(metadata.getCheckpointId());
+					resultFuture.completeExceptionally(ex);
+					throw ex;
+				}
+			},
+			"checkpoint %s with %s",
+			metadata,
+			options);
+		return resultFuture;
+	}
+
+	private void triggerSourcesCheckpoint(CheckpointBarrier checkpointBarrier) throws IOException {
+		for (StreamTaskSourceInput<?> sourceInput : operatorChain.getSourceTaskInputs()) {
```

Comment:
Shouldn't we differentiate for which `sourceInput` current barrier is (and call `processBarrier` only for it)?

Reply:
There is no need to do that. Please check the updated java doc in `StreamTaskSourceInput#checkpointStarted`.

Runtime has a flexibility to checkpoint the sources at any point of time, as long as it is in sync with network inputs.